Intel Mkl kernels for Concat, LRN and FusedBatchNorm

nhasabni opened the pull request:
This commit implements following changes:

1) MKL kernels for LRN, Concat, ConcatV2 and FusedBatchNorm and applicable gradient ops
2) Graph layout pass change to support new ops
3) Removes Mkl graph node merge pass as a separate pass
4) Changes MKL-TF tensor ordering from interleaved to contiguous

nhasabni committed:
Intel Mkl kernels for Concat, LRN and FusedBatchNorm

This commit implements following changes:

1) MKL kernels for LRN, Concat, ConcatV2 and FusedBatchNorm and applicable gradient ops
2) Graph layout pass change to support new ops
3) Removes Mkl graph node merge pass as a separate pass
4) Changes MKL-TF tensor ordering from interleaved to contiguous

tensorflow-jenkins commented:
Can one of the admins verify this patch?

zhangyaobit commented on tensorflow/core/kernels/mkl_fused_batch_norm_op.cc:
```diff
@@ -0,0 +1,688 @@
+/* Copyright 2016 The TensorFlow Authors. All Rights Reserved.
```
How is this tested? Is it possible to reuse the tests here? tensorflow/python/ops/nn_fused_batchnorm_test.py

nhasabni replied:
Yes, we used the same test for testing our kernel.

zhangyaobit commented on tensorflow/core/kernels/mkl_concat_op.cc:
```diff
@@ -0,0 +1,455 @@
+/* Copyright 2015 The TensorFlow Authors. All Rights Reserved.
```
Add tests?

nhasabni replied:
We used concat python test for this kernel.

zhangyaobit reviewed (commented):
Substantial and very impressive changes! I took a quick look at about half of the files; here are some initial comments.

zhangyaobit commented on tensorflow/core/graph/mkl_layout_pass.cc:
```diff
@@ -521,33 +685,36 @@ class MklLayoutRewritePass : public GraphOptimizationPass {
   static void CopyAttrsBiasAddGrad(const Node* orign, NodeBuilder* nb);
```
orgin?

nhasabni replied:
It meant original node. May be the name is not right. I changed it to orig_node.

zhangyaobit reviewed (commented):
Some initial comments. Mostly style and readability issues; lots of places need cleaning up.

zhangyaobit commented on tensorflow/core/graph/mkl_optimizer_merge.cc:
```diff
@@ -1,651 +0,0 @@
-/* Copyright 2015 The TensorFlow Authors. All Rights Reserved.
```
Are same functionalities moved to mkl_layout_pass.cc?

nhasabni replied:
Yes.

zhangyaobit commented on tensorflow/core/graph/mkl_layout_pass.cc:
```diff
@@ -118,10 +118,14 @@ namespace tensorflow {
 //      take (A, B, A_m, B_m) as input. Among N inputs one can get N!
```
This file is getting very long. Maybe separate all declarations to mkl_layout_pass.h, that would make it easier to read.

nhasabni replied:
Fixed

nhasabni replied:
Andy suggested earlier (for mkl_optimizer_merge review) to keep all the declarations that are not exported in C++, and only maintain a minimal interface in the header file. The header file is used in graph layout unit test so I have kept only the function required for unit test in the header file.

nhasabni replied:
Fixed

zhangyaobit commented on tensorflow/core/graph/mkl_tfconversion_pass.cc:
```diff
@@ -77,6 +77,7 @@ class MklToTfConversionPass : public GraphOptimizationPass {
   // false, otherwise.
```
Line 45, I think we can remove the usage of layer, which is kinds of confusing. We can consistently use node, and the an output node is usually referred as a target node or fetch node. The predecessor of TensorFlow, DistBelief, uses the concept of layer and TensorFlow avoided this concept on purpose and instead uses node and graph to represent more general-purpose computation structures.

nhasabni replied:
Fixed

zhangyaobit commented on tensorflow/core/graph/mkl_tfconversion_pass.cc:
```diff
@@ -77,6 +77,7 @@ class MklToTfConversionPass : public GraphOptimizationPass {
   // false, otherwise.
   bool RunPass(std::unique_ptr<Graph>* g);
```
Move all declarations of this file to the .h file?

nhasabni replied:
Pls refer to reply above.

zhangyaobit commented on tensorflow/core/graph/mkl_tfconversion_pass.cc:
```diff
@@ -106,7 +107,7 @@ const OptimizationPassRegistry::Grouping kMklTfConvPassGroup =
 
 Status MklToTfConversionPass::InsertConversionNodeOnEdge(
-    std::unique_ptr<Graph>* g, Edge* e) {
+    std::unique_ptr<Graph>* g, Edge *e) {
```
"Edge* e", to be consistent with "std::unique_ptr<Graph>*"?

zhangyaobit commented on tensorflow/core/graph/mkl_tfconversion_pass.cc:
```diff
@@ -180,7 +181,7 @@ bool MklToTfConversionPass::RunPass(std::unique_ptr<Graph>* g) {
   std::vector<Edge*> candidate_edges;
 
-  for (const Edge* e : (*g)->edges()) {
+  for (const Edge *e : (*g)->edges()) {
```
Edge*

agramesh1 committed:
Fixes based on feedback from Google on mkl-kernel-2

nhasabni commented on tensorflow/core/graph/mkl_layout_pass.cc:
```diff
@@ -118,10 +118,14 @@ namespace tensorflow {
+//      Tensorflow tensor immediately. In the context of above example, it
+//      will be: (A, A_m, B, B_m). We follow same ordering rule for output
+//      tensors. Contiguous ordering means all Tensorflow tensors are
+//      contiguous followd by all contiguous Mkl tensors. As a default
```
Fixed

nhasabni commented on tensorflow/core/graph/mkl_layout_pass.cc:
```diff
@@ -118,10 +118,14 @@ namespace tensorflow {
+//      will be: (A, A_m, B, B_m). We follow same ordering rule for output
+//      tensors. Contiguous ordering means all Tensorflow tensors are
+//      contiguous followd by all contiguous Mkl tensors. As a default
+//      ordering, we use Contiguous one.
```
Fixed

nhasabni commented on tensorflow/core/graph/mkl_layout_pass.cc:
```diff
@@ -208,8 +213,8 @@ namespace tensorflow {
 //        Backward op name = name of the op in the backward pass that receives
 //          workspace from forward op (MaxPoolGrad in the example)
 //        Slot = Number of the output or input slot that will be
```
Fixed

nhasabni commented on tensorflow/core/graph/mkl_layout_pass.cc:
```diff
@@ -234,7 +239,7 @@ namespace tensorflow {
 // -------------------------------------------
 // Consider BiasAddGrad op as:
 //
```
Fixed

nhasabni commented on tensorflow/core/graph/mkl_layout_pass.cc:
```diff
@@ -234,7 +239,7 @@ namespace tensorflow {
+//           O = MklConv2D(A, B, C, A_m, B_m, C_m)
 //           P = BiasAddGrad(O)
 //
 // Then we rewrite is as:
```
I fixed error in line 221. As per my understanding of English, I think line 227 is correct. Feel free to let me know if I am wrong.

nhasabni replied:
Fixed

zhangyaobit replied:
I think in this case, the period needs to be after the parenthesis :)

nhasabni replied:
ok, will change it.

nhasabni commented on tensorflow/core/graph/mkl_layout_pass.cc:
```diff
@@ -262,6 +267,10 @@ class MklLayoutRewritePass : public GraphOptimizationPass {
     csinfo_.biasadd           = "BiasAdd";
     csinfo_.matmul            = "MatMul";
     csinfo_.biasaddgrad       = "BiasAddGrad";
+
```
In line 247, we do not rewrite MklConv2D. We only rewrite BiasAddGrad into Conv2DWithBiasBackpropBias. MklConv2D is used in the example because that is the context that we look for while rewriting BiasAddGrad.

nhasabni replied:
Right, the distance in the example is 1.

nhasabni replied:
Fixed

nhasabni commented on tensorflow/core/graph/mkl_layout_pass.cc:
```diff
@@ -270,32 +279,53 @@ class MklLayoutRewritePass : public GraphOptimizationPass {
     csinfo_.avgpoolgrad       = "AvgPoolGrad";
     csinfo_.conv2dgradinput   = "Conv2DBackpropInput";
     csinfo_.conv2dgradfilter  = "Conv2DBackpropFilter";
```
Fixed

nhasabni commented on tensorflow/core/graph/mkl_layout_pass.cc:
```diff
@@ -270,32 +279,53 @@ class MklLayoutRewritePass : public GraphOptimizationPass {
+    csinfo_.fused_batch_norm = "FusedBatchNorm";
+    csinfo_.fused_batch_norm_grad = "FusedBatchNormGrad";
 
     rinfo_.push_back({csinfo_.conv2d,   csinfo_.mklconv2d,
```
Yes it is missing GetMklOpName.

It is not always the case that new name has any relationship with oldname. BiasAddGrad is an example. We are planning to add that op soon.

zhangyaobit replied:
I'm curious why BiasAddGrad is an exception? Thanks.

nhasabni commented on tensorflow/core/kernels/mkl_lrn_op.cc:
```diff
@@ -0,0 +1,775 @@
+/* Copyright 2015 The TensorFlow Authors. All Rights Reserved.
```
We used LRN Python test for testing the kernel. Graph rewrite pass rewrites them into our LRN kernel.

nhasabni commented on tensorflow/core/graph/mkl_layout_pass.cc:
```diff
@@ -322,13 +352,13 @@ class MklLayoutRewritePass : public GraphOptimizationPass {
-    int    numins;  // Number of inputs to the original op
+    string name;     // Original name of the op in the graph
+    string newname;  // New name of op in the graph
+    int numins;      // Number of inputs to the original op
```
Fixed

nhasabni commented on tensorflow/core/graph/mkl_layout_pass.cc:
```diff
@@ -322,13 +352,13 @@ class MklLayoutRewritePass : public GraphOptimizationPass {
+    int numins;      // Number of inputs to the original op
     // Function handler to copy attributes from old node to new node.
     std::function<void(const Node*, NodeBuilder*)> copyattrs;
     std::function<bool(const Node*)> rewriterule;  // Rule under which to
```
Fixed

nhasabni commented on tensorflow/core/graph/mkl_layout_pass.cc:
```diff
@@ -348,25 +378,28 @@ class MklLayoutRewritePass : public GraphOptimizationPass {
+    string node;    // Name of the node to be rewritten
+    string fwd;     // Node name in forward pass that this node
+                    // corresponds to
     size_t maxhop;  // Maximum number of hops the fwd is located
```
Fixed

nhasabni commented on tensorflow/core/graph/mkl_layout_pass.cc:
```diff
@@ -418,8 +455,43 @@ class MklLayoutRewritePass : public GraphOptimizationPass {
   // Clear all visited nodes
-  inline void UnMarkRewrittenNodes() {
-    visited_nodes_.clear();
+  inline void UnMarkRewrittenNodes() { visited_nodes_.clear(); }
```
Fixed

nhasabni replied:
Fixed

nhasabni commented on tensorflow/core/graph/mkl_layout_pass.cc:
```diff
@@ -455,7 +527,7 @@ class MklLayoutRewritePass : public GraphOptimizationPass {
   // We check for 2 scenarios for rewrite.
```
Fixed

nhasabni commented on tensorflow/core/graph/mkl_layout_pass.cc:
```diff
@@ -496,23 +568,115 @@ class MklLayoutRewritePass : public GraphOptimizationPass {
   //         Otherwise, it is not updated.
   Status RewriteNode(std::unique_ptr<Graph>* g, Node* n, const RewriteInfo* ri);
 
+  // Create a node that will feed a list of TF tensors to the new
```
Fixed

nhasabni commented on tensorflow/core/graph/mkl_layout_pass.cc:
```diff
@@ -496,23 +568,115 @@ class MklLayoutRewritePass : public GraphOptimizationPass {
+    int* input_idx, int list_length,
+    std::vector<NodeBuilder::NodeOut>* output_nodes);
+
+  // Create a node that will feed a list of Mkl tensors to the new
```
Fixed

nhasabni commented on tensorflow/core/graph/mkl_layout_pass.cc:
```diff
@@ -496,23 +568,115 @@ class MklLayoutRewritePass : public GraphOptimizationPass {
+    int* input_idx, int list_length,
+    std::vector<NodeBuilder::NodeOut>* output_nodes);
+
+  // Create a node that will feed an Mkl tensor to the new
```
Fixed

nhasabni commented on tensorflow/core/graph/mkl_layout_pass.cc:
```diff
@@ -496,23 +568,115 @@ class MklLayoutRewritePass : public GraphOptimizationPass {
+    std::vector<NodeBuilder::NodeOut>* workspace_tensors,
+    bool are_workspace_tensors_available);
+
+#if 0
```
Removed

nhasabni commented on tensorflow/core/graph/mkl_layout_pass.cc:
```diff
@@ -565,40 +732,21 @@ static void FillInputs(const Node* n,
-    const gtl::InlinedVector<std::pair<Node*, int>, 4>& inputs, NodeBuilder* nb,
-    Node* orign) {
-  std::vector<NodeBuilder::NodeOut> new_inputs;
+#if 0
```
I have removed code section under #if 0

nhasabni committed:
Fixes to Mkl graph passes based on review comments

drpngx commented:
@nhasabni could you pull rebase and push again? googlebot seems confused.

@zhangyaobit how do the new changes look?

nhasabni committed:
Updates based on review comments - round 2

This commit makes following changes based on review comments:
1) Changes API that uses notion of Mkl layer to use Mkl Op
2) Adds description of workspace
3) Fixes incorrectly updated mkl_relu

vivek-rane commented:
@zhangyaobit @drpngx We have updated the code based on the review feedback. I had to reset the checkins to fix the CLA issue. Please can you continue reviewing and testing?

drpngx commented:
Yes, @zhangyaobit I think it's ready for another pass.

Jenkins, test this please.

vivek-rane commented:
Tests are green :)

zhangyaobit commented on tensorflow/core/graph/mkl_layout_pass.cc:
```diff
@@ -191,9 +195,10 @@ namespace tensorflow {
 //        B = MaxPoolGrad(X, A, Y)
 //
 //        We will transform this graph to propagate workspace as:
+//        (with contiguous ordering)
```
Did/could you add the definition of workspace, when this term is used for the first time?

nhasabni replied:
Yes, I did add it - https://github.com/tensorflow/tensorflow/pull/8968/commits/67f9925ef9ceed02892c200a3122092ab497943a#diff-2359860d8d805c348b1e2af4598b43dd

nhasabni commented on tensorflow/core/kernels/mkl_fused_batch_norm_op.cc:
```diff
@@ -0,0 +1,688 @@
+                    mkl_context.mkl_res_batchnorm[dnnResourceVariance]),
+                scale.NumElements() * sizeof(float));
+
+    // Bessel's correction on variance, if training mode in on
```
Fixed

nhasabni commented:
@zhangyaobit Actually the example at line 248 explains "within context matching depth". To put an example for "outside context matching depth" we would need to add at least 10 lines of example. We think that may not be desirable. We have explained what the context and the depth means, so we think that should be enough. Also, there are unit tests for context match.

For renaming, Conv2DWithBiasBackpropBias to Conv2DWithBiasGrad, we used naming consistent with TensorFlow's naming convention. We found Conv2DBackpropFilter and Conv2DBackpropInput. Since we merge Conv2D and Bias together, we have Conv2DWithBiasBackpropBias. We don't think Conv2DWithBiasGrad is intuitive here.

zhangyaobit commented on tensorflow/core/graph/mkl_layout_pass.cc:
```diff
@@ -561,76 +727,273 @@ static void FillInputs(const Node* n,
+  // Actual number of inputs can be greater than or equal to number
+  // of Input slots because inputs of type list could be unfolded.
+  CHECK_GE(old_node_inputs.size(), old_node_input_slots);
+  int nnsidx = 0;  // slot index for inputs of new node
```
nnsidx -> nn_slot_idx?

zhangyaobit commented on tensorflow/core/graph/mkl_layout_pass.cc:
```diff
@@ -561,76 +727,273 @@ static void FillInputs(const Node* n,
+  int nnsidx = 0;  // slot index for inputs of new node
+
+  // Let's copy all inputs (TF tensors) of original node to new node.
+  int iidx = 0;
```
what does the first i mean here? iidx->i_idx?

nhasabni replied:
input index

zhangyaobit commented on tensorflow/core/graph/mkl_layout_pass.cc:
```diff
@@ -561,76 +727,273 @@ static void FillInputs(const Node* n,
+
+  // Let's copy all inputs (TF tensors) of original node to new node.
+  int iidx = 0;
+  for (int onsidx = 0; onsidx < old_node_input_slots; onsidx++) {
```
onsidx->on_slot_idx?

zhangyaobit commented on tensorflow/core/graph/mkl_layout_pass.cc:
```diff
@@ -654,39 +1017,41 @@ void MklLayoutRewritePass::GetDummyWorkspaceTensorNode(
+                  .Attr("value", proto)
+                  .Attr("dtype", dt)
+                  .Device(orig_node->def().device())  // We place this node on
+                  // same device as device of original node.
```
Fix indentation?

zhangyaobit commented on tensorflow/core/graph/mkl_layout_pass.cc:
```diff
@@ -1049,81 +1536,84 @@ Status MklLayoutRewritePass::RewriteNode(std::unique_ptr<Graph>* g, Node* orign,
+  nb.Attr("_kernel", mkl_op_registry::kMklOpLabel);
 
   // Finalize graph and get new node.
   Node* newn = nullptr;
```
newn->new_node?

zhangyaobit reviewed (commented):
Mostly issues to improve clarity and readability of the comments and code.

zhangyaobit commented on tensorflow/core/graph/mkl_tfconversion_pass.cc:
```diff
@@ -123,22 +124,22 @@ Status MklToTfConversionPass::InsertConversionNodeOnEdge(
   if (src_datatype != dst_datatype) {
-    string err_msg = "T attribute of " + src->name() + " and " + dst->name() +
-                     " do not match. Will not insert" +
+    string err_msg = "T attribute of " + src->name() + " and " +
```
Line 94: "edge will be deleted ...", could you point to me in the code where the deletion of edge is?

nhasabni replied:
line 166: (*g)->RemoveEdge(e)

nhasabni commented on tensorflow/core/graph/mkl_layout_pass.cc:
```diff
@@ -322,69 +366,74 @@ class MklLayoutRewritePass : public GraphOptimizationPass {
   /// the number of inputs to the original op, and the function to be used
```
Fixed

nhasabni commented on tensorflow/core/graph/mkl_layout_pass.cc:
```diff
@@ -322,69 +366,74 @@ class MklLayoutRewritePass : public GraphOptimizationPass {
-    string name;     // Original name of the op in the graph
-    string newname;  // New name of op in the graph
-    int numins;      // Number of inputs to the original op
+    string name;      // Original name of the op in the graph
```
Fixed

nhasabni commented on tensorflow/core/graph/mkl_tfconversion_pass.cc:
```diff
@@ -77,14 +77,15 @@ class MklToTfConversionPass : public GraphOptimizationPass {
   // false, otherwise.
```
Fixed

nhasabni commented on tensorflow/core/graph/mkl_tfconversion_pass.cc:
```diff
@@ -210,7 +211,6 @@ bool MklToTfConversionPass::RunPass(std::unique_ptr<Graph>* g) {
 
     // Check if src with is Mkl-compliant, while dst is not Mkl-compliant.
-
     if (IsMklSupportedOp(src->type_string(), src_datatype) &&
```
Fixed

nhasabni commented:
@zhangyaobit I've addressed your comments so far. I will commit my changes if you are done with your review. Otherwise, I will wait.

zhangyaobit commented on tensorflow/core/kernels/mkl_concat_op.cc:
```diff
@@ -0,0 +1,455 @@
+  typedef std::vector<std::unique_ptr<typename TTypes<T, 2>::ConstMatrix>>
+      ConstMatrixVector;
+
+  explicit EigenConcatBaseOp(OpKernelConstruction* c) : OpKernel(c) {}
```
Is this code same with the one in core/kernels/concat_op.cc? Are you able to make concat_op.h for concat_op.cc and use the existing ConcatBaseOp?

nhasabni replied:
No, the code is not exactly same. EigenConcatBaseOp takes input tensors differently than original Eigen version.

zhangyaobit replied:
Could you modify the input with OpKernelContext::mutable_input, so that ConcatBaseO could be reused? https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/framework/op_kernel.h#L632

zhangyaobit commented on tensorflow/core/kernels/mkl_concat_op.cc:
```diff
@@ -0,0 +1,455 @@
+    }
+
+    // Check that total number of dimensions is 4, if not call Eigen.
+    if (!invoke_eigen) {
```
Could/Should this decision be made during graph rewrite, instead of inside the kernel?

nhasabni replied:
No, the decision cannot be made in graph rewrite pass. We cannot find out concat_dim in graph rewrite. Also, we do not know if input tensors are in Mkl layout or Tensorflow layout in graph rewrite.

zhangyaobit commented on tensorflow/core/kernels/mkl_lrn_op.cc:
```diff
@@ -0,0 +1,775 @@
+      lrn_res[dnnResourceWorkspace] = mkl_buf_workspace;
+    }
+
+    // Fallback implementation - Taken from lrn_op.cc
```
Could you make lrn_op.h and expose this function, so that we could reuse this code instead of making a copy?

zhangyaobit commented on tensorflow/core/kernels/mkl_lrn_op.cc:
```diff
@@ -0,0 +1,775 @@
+// Create a depth-by-depth band matrix with 1s along a swath of size (2 *
+// depth_radius + 1) around the diagonal.
+template <typename T>
+void GetBandMatrix(int depth, int depth_radius,
```
Could reuse the one in lrn_op.cc by creating lrn_op.h

zhangyaobit commented on tensorflow/core/kernels/mkl_lrn_op.cc:
```diff
@@ -0,0 +1,775 @@
+        mkl_context.lt_input =
+            static_cast<dnnLayout_t>(mkl_context.input_shape.GetCurLayout());
+        workspace_enabled_ = true;
+        /* std::cout<<" the mkl lrn is enabled!!!!!!!!!!"<<std::endl; */
```
Clean up? (And a few other places)

zhangyaobit commented on tensorflow/core/kernels/mkl_fused_batch_norm_op.cc:
```diff
@@ -0,0 +1,688 @@
+    void MklCleanup() {
+      bool input_in_mkl_format = mkl_shape_input_shape.IsMklTensor();
+      if (!input_in_mkl_format) dnnLayoutDelete_F32(mkl_lt_input);
+
```
Delete empty line?

zhangyaobit commented on tensorflow/core/ops/nn_ops.cc:
```diff
@@ -2843,6 +2843,211 @@ NOTE Do not invoke this operator directly in Python. Graph rewrite pass is
+    .Output("y: T")
+    .Output("batch_mean: T")
+    .Output("batch_variance: T")
+    .Output("reserve_space_1: T")
```
Looks are reserve_space_1 and reserve_space_2 are needed in this case. Are you able to reuse batch_mean and batch_variance in MklFusedBatchNormGrad? If so, I think we can remove these two reserved space? 

For CuDNN, this is not the case, because they compute the inverse of variance, which is not compatible with TensorFlow, and that's why we introduce two two reserved spaces; but for Mkl, looks like these two reserved spaces are not needed.

nhasabni replied:
One of my team members will address these in the next version.

zhangyaobit commented on tensorflow/core/ops/nn_ops.cc:
```diff
@@ -2843,6 +2843,211 @@ NOTE Do not invoke this operator directly in Python. Graph rewrite pass is
+    .Input("x: T")
+    .Input("scale: T")
+    .Input("reserve_space_1: T")
+    .Input("reserve_space_2: T")
```
If reserved_space_{1,2} are removed, these two could be renamed to batch_mean, and batch_var.

nhasabni replied:
One of my team members will address these in the next version.

zhangyaobit commented on tensorflow/core/ops/nn_ops.cc:
```diff
@@ -2843,6 +2843,211 @@ NOTE Do not invoke this operator directly in Python. Graph rewrite pass is
+    .Output("mkl_x_backprop: uint8")
+    .Output("mkl_scale_backprop: uint8")
+    .Output("mkl_offset_backprop: uint8")
+    .Output("mkl_reserve_space_3: uint8")
```
Maybe remove these two reserve spaces as well. I think they are only needed if to be used a Python interface. But this is C++ only and handled during graph rewrite, so I guess they are not needed here.

nhasabni replied:
One of my team members will address these in the next version.

rmlarsen commented:
@nhasabni @zhangyaobit could we address the remaining style and minor corrections now and maybe defer major refactoring for better code re-use to a followup?

zhangyaobit commented:
@nhasabni @rmlarsen, sounds good to me. Please add TODOs if you prefer to address  later.

nhasabni commented:
@rmlarsen Yes, sounds good. We have addressed comments for graph passes. We will ensure to add TODOs to kernels and address style/minor correction issues now.

jinghuangintel committed:
updated based on review comments for mkl_lrn

nhasabni committed:
Updates based on review comments - round 3

zhangyaobit commented on tensorflow/core/graph/mkl_layout_pass.cc:
```diff
@@ -87,106 +87,104 @@ namespace tensorflow {
-// represents a dummy Mkl tensor.
+// When a previous node in the graph is an Mkl node, A_m will represent a valid
+// Mkl tensor. But when a previous node is not an Mkl node, A_m will represent
+// represent a dummy Mkl tensor.
```
Delete the second "represent"

zhangyaobit commented on tensorflow/core/graph/mkl_layout_pass.cc:
```diff
@@ -87,106 +87,104 @@ namespace tensorflow {
-//      take (A, B, A_m, B_m) as input. Among N inputs one can get N!
-//      permutations.
+//      Since every rewritten node generates twice the number of inputs and
+//      outputs, one ould imagine various orderings among Tensorflow tensors
```
ould --> could

zhangyaobit reviewed (commented):
Looks good! Thanks!

nhasabni committed:
Minor English corrections in mkl layout pass

nhasabni committed:
Addressing some of review comments in MklFusedBatchNorm

zhangyaobit reviewed (approved):
Thank you all for this significant contribution! Look forward to great performance speedups!

zhangyaobit commented:
Jenkins, test this please.
